From bdd594e52a508f4e41b2c726c951e5d706ce0ad6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Oliver=20St=C3=A4bler?= <oliver.staebler@bytesatwork.ch>
Date: Wed, 4 Feb 2015 17:30:30 +0100
Subject: [PATCH] bytePANEL: Enable CAN

* Enable C_CAN in defconfig
* Enable and mux in device tree
* Set GPIOs to enable CAN on bytepanel in board file
---
 arch/arm/boot/dts/bytepanel.dtsi     | 19 +++++++++++++++++--
 arch/arm/configs/bytepanel_defconfig | 36 ++++++++++++++++++++++++++++++++++--
 arch/arm/mach-omap2/io.c             | 13 +++++++++++++
 3 files changed, 64 insertions(+), 4 deletions(-)

diff --git a/arch/arm/boot/dts/bytepanel.dtsi b/arch/arm/boot/dts/bytepanel.dtsi
index a49c306..e797b98 100644
--- a/arch/arm/boot/dts/bytepanel.dtsi
+++ b/arch/arm/boot/dts/bytepanel.dtsi
@@ -43,8 +43,9 @@
 		pinctrl-single,pins = <
 			PIN_SPI0_CS0		0
 			PIN_SPI0_D1		0
-			PIN_UART1_TXD		0
-			PIN_UART1_RXD		0
+			PIN_MCASP0_FSR		(MUX_MODE7 | PIN_OUTPUT_PULLDOWN)
+			PIN_MCASP0_ACLKX	(MUX_MODE7 | PIN_OUTPUT_PULLDOWN)
+			PIN_MCASP0_ACLKR	(MUX_MODE7 | PIN_OUTPUT_PULLDOWN)
 		>;
 	};
 
@@ -158,6 +159,13 @@
 			PIN_MCASP0_AHCLKR	(MUX_MODE2 | PIN_INPUT_PULLDOWN)	// axr2
 		>;
 	};
+
+	dcan1_pins: dcan1_pins {
+		pinctrl-single,pins = <
+			PIN_UART1_RXD		(MUX_MODE2 | PIN_OUTPUT)		// dcan1_tx
+			PIN_UART1_TXD		(MUX_MODE2 | PIN_INPUT_PULLDOWN)	// dcan1_rx
+		>;
+	};
 };
 
 // ---- UART0 on X3 ----
@@ -275,3 +283,10 @@
 	tx-num-evt = <1>;
 	rx-num-evt = <1>;
 };
+
+&dcan1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&dcan1_pins>;
+
+	status = "okay";
+};
diff --git a/arch/arm/configs/bytepanel_defconfig b/arch/arm/configs/bytepanel_defconfig
index 3f47453..8158f87 100644
--- a/arch/arm/configs/bytepanel_defconfig
+++ b/arch/arm/configs/bytepanel_defconfig
@@ -736,7 +736,39 @@ CONFIG_BQL=y
 # CONFIG_NET_TCPPROBE is not set
 # CONFIG_NET_DROP_MONITOR is not set
 # CONFIG_HAMRADIO is not set
-# CONFIG_CAN is not set
+CONFIG_CAN=y
+CONFIG_CAN_RAW=y
+CONFIG_CAN_BCM=y
+CONFIG_CAN_GW=y
+
+#
+# CAN Device Drivers
+#
+# CONFIG_CAN_VCAN is not set
+# CONFIG_CAN_SLCAN is not set
+CONFIG_CAN_DEV=y
+CONFIG_CAN_CALC_BITTIMING=y
+# CONFIG_CAN_LEDS is not set
+# CONFIG_CAN_AT91 is not set
+# CONFIG_CAN_TI_HECC is not set
+# CONFIG_CAN_MCP251X is not set
+# CONFIG_CAN_FLEXCAN is not set
+# CONFIG_CAN_GRCAN is not set
+# CONFIG_CAN_SJA1000 is not set
+CONFIG_CAN_C_CAN=y
+CONFIG_CAN_C_CAN_PLATFORM=y
+# CONFIG_CAN_CC770 is not set
+
+#
+# CAN USB interfaces
+#
+# CONFIG_CAN_EMS_USB is not set
+# CONFIG_CAN_ESD_USB2 is not set
+# CONFIG_CAN_KVASER_USB is not set
+# CONFIG_CAN_PEAK_USB is not set
+# CONFIG_CAN_8DEV_USB is not set
+# CONFIG_CAN_SOFTING is not set
+# CONFIG_CAN_DEBUG_DEVICES is not set
 # CONFIG_IRDA is not set
 # CONFIG_BT is not set
 # CONFIG_AF_RXRPC is not set
@@ -1755,7 +1787,7 @@ CONFIG_MFD_CORE=y
 # CONFIG_MFD_SMSC is not set
 # CONFIG_ABX500_CORE is not set
 # CONFIG_MFD_STMPE is not set
-# CONFIG_MFD_SYSCON is not set
+CONFIG_MFD_SYSCON=y
 # CONFIG_MFD_TI_AM335X_TSCADC is not set
 # CONFIG_MFD_LP3943 is not set
 # CONFIG_MFD_LP8788 is not set
diff --git a/arch/arm/mach-omap2/io.c b/arch/arm/mach-omap2/io.c
index a122649..d773635 100644
--- a/arch/arm/mach-omap2/io.c
+++ b/arch/arm/mach-omap2/io.c
@@ -21,6 +21,7 @@
 #include <linux/init.h>
 #include <linux/io.h>
 #include <linux/clk.h>
+#include <linux/gpio.h>
 
 #include <asm/tlb.h>
 #include <asm/mach/map.h>
@@ -588,11 +589,23 @@ void __init am33xx_init_early(void)
 	omap_clk_soc_init = am33xx_dt_clk_init;
 }
 
+#define GPIO_MULTIPLEXER_A0 115
+#define GPIO_MULTIPLEXER_A1 110
+#define GPIO_MULTIPLEXER_A2 114
+
+void __init bytepanel_hacks(void)
+{
+	gpio_request_one(GPIO_MULTIPLEXER_A0, GPIOF_OUT_INIT_HIGH | GPIOF_EXPORT, "gpio_multiplexer_a0");
+	gpio_request_one(GPIO_MULTIPLEXER_A1, GPIOF_OUT_INIT_HIGH | GPIOF_EXPORT, "gpio_multiplexer_a1");
+	gpio_request_one(GPIO_MULTIPLEXER_A2, GPIOF_OUT_INIT_LOW | GPIOF_EXPORT, "gpio_multiplexer_a2");
+}
+
 void __init am33xx_init_late(void)
 {
 	am33xx_opp_init();
 	omap_common_late_init();
 	am33xx_pm_init();
+	bytepanel_hacks();
 }
 #endif
 
-- 
2.0.5

